File: backend/src/main/java/com/europace/bowling/service/BowlingScoreService.java
package com.europace.bowling.service;

import com.europace.bowling.dto.FrameResult;
import com.europace.bowling.dto.ScoreResponse;
import com.europace.bowling.exception.InvalidRollException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

@Service
public class BowlingScoreService {

    private static final Logger log = LoggerFactory.getLogger(BowlingScoreService.class);
    private static final int MAX_PINS = 10;
    private static final int MAX_FRAMES = 10;

    public ScoreResponse calculateScore(List<Integer> rolls) {
        log.debug("Calculating score for {} rolls", rolls.size());
        validateRolls(rolls);

        List<FrameResult> frames = buildFrames(rolls);
        int totalScore = frames.stream()
            .filter(f -> f.cumulativeScore() != null)
            .mapToInt(FrameResult::cumulativeScore)
            .max().orElse(0);
        boolean gameOver = !frames.isEmpty() && frames.get(MAX_FRAMES - 1).isComplete();
        int currentFrame = findCurrentFrame(frames);
        int pinsRemaining = gameOver ? 0 : calculatePinsRemaining(frames, currentFrame);

        return new ScoreResponse(totalScore, currentFrame, pinsRemaining, gameOver, frames);
    }

    private void validateRolls(List<Integer> rolls) {
        for (Integer pins : rolls) {
            if (pins == null || pins < 0 || pins > MAX_PINS) {
                throw new InvalidRollException("Pin count must be between 0 and " + MAX_PINS);
            }
        }
    }

    private List<FrameResult> buildFrames(List<Integer> rolls) {
        List<FrameResult> frames = new ArrayList<>();
        int rollIndex = 0;

        for (int frameIndex = 0; frameIndex < MAX_FRAMES; frameIndex++) {
            boolean isTenth = frameIndex == MAX_FRAMES - 1;

            if (isTenth) {
                frames.add(buildTenthFrame(rolls, rollIndex, frames));
                break;
            }

            if (rollIndex >= rolls.size()) {
                frames.add(emptyFrame(false));
                continue;
            }

            int first = rolls.get(rollIndex);

            if (first == MAX_PINS) {
                boolean hasBonus = rolls.size() > rollIndex + 2;
                Integer score = hasBonus
                    ? MAX_PINS + rolls.get(rollIndex + 1) + rolls.get(rollIndex + 2)
                    : null;
                Integer prev = prevCumulative(frames);
                frames.add(new FrameResult(
                    List.of(first), score,
                    score != null && prev != null ? prev + score : null,
                    true, false, true, false
                ));
                rollIndex += 1;
                continue;
            }

            if (rolls.size() <= rollIndex + 1) {
                frames.add(new FrameResult(
                    List.of(first), null, null,
                    false, false, false, false
                ));
                rollIndex += 1;
                continue;
            }

            int second = rolls.get(rollIndex + 1);
            boolean isSpare = first + second == MAX_PINS;
            Integer prev = prevCumulative(frames);

            if (isSpare) {
                boolean hasBonus = rolls.size() > rollIndex + 2;
                Integer score = hasBonus ? MAX_PINS + rolls.get(rollIndex + 2) : null;
                frames.add(new FrameResult(
                    List.of(first, second), score,
                    score != null && prev != null ? prev + score : null,
                    false, true, true, false
                ));
            } else {
                int score = first + second;
                frames.add(new FrameResult(
                    List.of(first, second), score,
                    prev != null ? prev + score : score,
                    false, false, true, false
                ));
            }
            rollIndex += 2;
        }

        while (frames.size() < MAX_FRAMES) {
            frames.add(emptyFrame(frames.size() == MAX_FRAMES - 1));
        }

        return frames;
    }

    private FrameResult buildTenthFrame(List<Integer> rolls, int rollIndex, List<FrameResult> prev) {
        List<Integer> tenth = new ArrayList<>(
            rolls.subList(Math.min(rollIndex, rolls.size()), rolls.size())
        );
        Integer prevCum = prevCumulative(prev);

        if (tenth.isEmpty()) return emptyFrame(true);

        int r1 = tenth.get(0);
        boolean firstStrike = r1 == MAX_PINS;
        boolean spare = !firstStrike && tenth.size() >= 2 && r1 + tenth.get(1) == MAX_PINS;

        if (firstStrike) {
            if (tenth.size() < 3) return new FrameResult(new ArrayList<>(tenth), null, null, true, false, false, true);
            int score = tenth.get(0) + tenth.get(1) + tenth.get(2);
            return new FrameResult(tenth.subList(0, 3), score,
                prevCum != null ? prevCum + score : score, true, false, true, true);
        }

        if (tenth.size() < 2) return new FrameResult(new ArrayList<>(tenth), null, null, false, false, false, true);

        if (spare) {
            if (tenth.size() < 3) return new FrameResult(new ArrayList<>(tenth), null, null, false, true, false, true);
            int score = tenth.get(0) + tenth.get(1) + tenth.get(2);
            return new FrameResult(tenth.subList(0, 3), score,
                prevCum != null ? prevCum + score : score, false, true, true, true);
        }

        int score = tenth.get(0) + tenth.get(1);
        return new FrameResult(List.of(tenth.get(0), tenth.get(1)), score,
            prevCum != null ? prevCum + score : score, false, false, true, true);
    }

    private int findCurrentFrame(List<FrameResult> frames) {
        for (int i = 0; i < frames.size(); i++) {
            if (!frames.get(i).isComplete()) return i;
        }
        return MAX_FRAMES - 1;
    }

    private int calculatePinsRemaining(List<FrameResult> frames, int currentFrame) {
        FrameResult current = frames.get(currentFrame);
        if (currentFrame == 9) return tenthFramePins(current.rolls());
        if (current.rolls().isEmpty()) return MAX_PINS;
        return MAX_PINS - current.rolls().get(0);
    }

    private int tenthFramePins(List<Integer> rolls) {
        if (rolls.isEmpty()) return MAX_PINS;
        int r1 = rolls.get(0);
        if (rolls.size() == 1) return r1 == MAX_PINS ? MAX_PINS : MAX_PINS - r1;
        if (rolls.size() == 2) {
            if (r1 == MAX_PINS) {
                int r2 = rolls.get(1);
                return r2 == MAX_PINS ? MAX_PINS : MAX_PINS - r2;
            }
            return MAX_PINS;
        }
        return 0;
    }

    private Integer prevCumulative(List<FrameResult> frames) {
        if (frames.isEmpty()) return 0;
        return frames.get(frames.size() - 1).cumulativeScore();
    }

    private FrameResult emptyFrame(boolean isTenth) {
        return new FrameResult(Collections.emptyList(), null, null, false, false, false, isTenth);
    }
}

File: src/app/services/bowling-score.service.ts
import { computed, Injectable, signal } from '@angular/core';
import { Frame } from '../models/bowling.model';

@Injectable({ providedIn: 'root' })
export class BowlingScoreService {
  private readonly MAX_PINS = 10;
  private readonly MAX_FRAMES = 10;

  private readonly _rolls = signal<number[]>([]);

  readonly frames = computed(() => this.buildFrames(this._rolls()));

  readonly totalScore = computed(() => {
    const completed = this.frames().filter(f => f.cumulativeScore !== null);
    return completed.length > 0
      ? (completed[completed.length - 1].cumulativeScore ?? 0)
      : 0;
  });

  readonly isGameOver = computed(() => {
    const allFrames = this.frames();
    return allFrames.length === this.MAX_FRAMES &&
      allFrames[this.MAX_FRAMES - 1].isComplete;
  });

  readonly currentFrameIndex = computed(() => {
    const idx = this.frames().findIndex(f => !f.isComplete);
    return idx === -1 ? this.MAX_FRAMES - 1 : idx;
  });

  readonly pinsRemaining = computed(() =>
    this.calculatePinsRemaining(this._rolls())
  );

  addRoll(pins: number): void {
    if (this.isGameOver()) return;
    if (!this.isValidRoll(pins, this._rolls())) return;
    this._rolls.update(rolls => [...rolls, pins]);
  }

  resetGame(): void {
    this._rolls.set([]);
  }

  private isValidRoll(pins: number, rolls: number[]): boolean {
    if (!Number.isInteger(pins) || pins < 0 || pins > this.MAX_PINS) return false;
    return pins <= this.calculatePinsRemaining(rolls);
  }

  private buildFrames(rolls: number[]): Frame[] {
    const frames: Frame[] = [];
    let rollIndex = 0;

    for (let frameIndex = 0; frameIndex < this.MAX_FRAMES; frameIndex++) {
      const isTenth = frameIndex === this.MAX_FRAMES - 1;

      if (isTenth) {
        frames.push(this.buildTenthFrame(rolls, rollIndex, frames));
        break;
      }

      if (rollIndex >= rolls.length) {
        frames.push(this.emptyFrame(isTenth));
        continue;
      }

      const first = rolls[rollIndex];

      if (first === this.MAX_PINS) {
        const hasBonus = rolls.length > rollIndex + 2;
        const score = hasBonus
          ? this.MAX_PINS + rolls[rollIndex + 1] + rolls[rollIndex + 2]
          : null;
        const prev = this.prevCumulative(frames);
        frames.push({
          rolls: [first], score,
          cumulativeScore: score !== null && prev !== null ? prev + score : null,
          isStrike: true, isSpare: false, isComplete: true, isTenth
        });
        rollIndex += 1;
        continue;
      }

      if (rolls.length <= rollIndex + 1) {
        frames.push({
          rolls: [first], score: null, cumulativeScore: null,
          isStrike: false, isSpare: false, isComplete: false, isTenth
        });
        rollIndex += 1;
        continue;
      }

      const second = rolls[rollIndex + 1];
      const isSpare = first + second === this.MAX_PINS;
      const prev = this.prevCumulative(frames);

      if (isSpare) {
        const hasBonus = rolls.length > rollIndex + 2;
        const score = hasBonus ? this.MAX_PINS + rolls[rollIndex + 2] : null;
        frames.push({
          rolls: [first, second], score,
          cumulativeScore: score !== null && prev !== null ? prev + score : null,
          isStrike: false, isSpare: true, isComplete: true, isTenth
        });
      } else {
        const score = first + second;
        frames.push({
          rolls: [first, second], score,
          cumulativeScore: prev !== null ? prev + score : score,
          isStrike: false, isSpare: false, isComplete: true, isTenth
        });
      }
      rollIndex += 2;
    }

    while (frames.length < this.MAX_FRAMES) {
      frames.push(this.emptyFrame(frames.length === this.MAX_FRAMES - 1));
    }

    return frames;
  }

  private buildTenthFrame(rolls: number[], rollIndex: number, previousFrames: Frame[]): Frame {
    const tenthRolls = rolls.slice(rollIndex);
    const prev = this.prevCumulative(previousFrames);
    if (tenthRolls.length === 0) return this.emptyFrame(true);

    const [r1, r2, r3] = tenthRolls;
    const firstIsStrike = r1 === this.MAX_PINS;
    const isSpare = !firstIsStrike && r2 !== undefined && r1 + r2 === this.MAX_PINS;

    if (firstIsStrike) {
      if (tenthRolls.length < 3) {
        return { rolls: tenthRolls, score: null, cumulativeScore: null,
          isStrike: true, isSpare: false, isComplete: false, isTenth: true };
      }
      const score = r1 + r2 + r3;
      return { rolls: tenthRolls.slice(0, 3), score,
        cumulativeScore: prev !== null ? prev + score : score,
        isStrike: true, isSpare: false, isComplete: true, isTenth: true };
    }

    if (tenthRolls.length < 2) {
      return { rolls: tenthRolls, score: null, cumulativeScore: null,
        isStrike: false, isSpare: false, isComplete: false, isTenth: true };
    }

    if (isSpare) {
      if (tenthRolls.length < 3) {
        return { rolls: tenthRolls, score: null, cumulativeScore: null,
          isStrike: false, isSpare: true, isComplete: false, isTenth: true };
      }
      const score = r1 + r2 + r3;
      return { rolls: tenthRolls.slice(0, 3), score,
        cumulativeScore: prev !== null ? prev + score : score,
        isStrike: false, isSpare: true, isComplete: true, isTenth: true };
    }

    const score = r1 + r2;
    return { rolls: [r1, r2], score,
      cumulativeScore: prev !== null ? prev + score : score,
      isStrike: false, isSpare: false, isComplete: true, isTenth: true };
  }

  private calculatePinsRemaining(rolls: number[]): number {
    if (this.checkGameOver(rolls)) return 0;
    const frames = this.buildFrames(rolls);
    const currentIdx = frames.findIndex(f => !f.isComplete);
    if (currentIdx === -1) return 0;
    const current = frames[currentIdx];
    if (currentIdx === 9) return this.tenthFramePinsRemaining(current.rolls);
    return current.rolls.length === 0 ? this.MAX_PINS : this.MAX_PINS - current.rolls[0];
  }

  private tenthFramePinsRemaining(tenthRolls: number[]): number {
    if (tenthRolls.length === 0) return this.MAX_PINS;
    const [r1, r2] = tenthRolls;
    if (tenthRolls.length === 1) return r1 === this.MAX_PINS ? this.MAX_PINS : this.MAX_PINS - r1;
    if (tenthRolls.length === 2) {
      if (r1 === this.MAX_PINS) return r2 === this.MAX_PINS ? this.MAX_PINS : this.MAX_PINS - r2;
      return this.MAX_PINS;
    }
    return 0;
  }

  private checkGameOver(rolls: number[]): boolean {
    const frames = this.buildFrames(rolls);
    return frames.length === this.MAX_FRAMES && frames[this.MAX_FRAMES - 1].isComplete;
  }

  private prevCumulative(frames: Frame[]): number | null {
    return frames.length === 0 ? 0 : frames[frames.length - 1].cumulativeScore;
  }

  private emptyFrame(isTenth: boolean): Frame {
    return { rolls: [], score: null, cumulativeScore: null,
      isStrike: false, isSpare: false, isComplete: false, isTenth };
  }
}